src/hotspot/share/memory/virtualspace.cpp

```diff
@@ -115,30 +115,82 @@ static void unmap_or_release_memory(char* base, size_t size, bool is_file_mapped
     log_debug(gc, heap, coops)("Reserved memory not at requested address: " PTR_FORMAT " vs " PTR_FORMAT, p2i(base), p2i(requested_address));
-    // OS ignored requested address. Try different address.
-    if (special) {
-      if (!os::release_memory_special(base, size)) {
-        fatal("os::release_memory_special failed");
-      }
-    } else {
-      unmap_or_release_memory(base, size, is_file_mapped);
-    }
   }
   return true;
 }
 
+static bool use_explicit_large_pages(bool large) {
+  return !os::can_commit_large_page_memory() && large;
+}
+
+static bool large_pages_requested() {
+  return UseLargePages &&
+         (!FLAG_IS_DEFAULT(UseLargePages) || !FLAG_IS_DEFAULT(LargePageSizeInBytes));
+}
+
+static char* reserve_memory(char* requested_address, const size_t size,
+                            const size_t alignment, int fd, bool exec) {
+  char* base;
+  // If the memory was requested at a particular address, use
+  // os::attempt_reserve_memory_at() to avoid mapping over something
+  // important.  If the reservation fails, return NULL.
+  if (requested_address != 0) {
+    assert(is_aligned(requested_address, alignment),
+           "Requested address " PTR_FORMAT " must be aligned to " SIZE_FORMAT,
+           p2i(requested_address), alignment);
+    base = attempt_map_or_reserve_memory_at(requested_address, size, fd, exec);
+  } else {
+    // Optimistically assume that the OS returns an aligned base pointer.
+    // When reserving a large address range, most OSes seem to align to at
+    // least 64K.
+    base = map_or_reserve_memory(size, fd, exec);
+    // Check alignment constraints. This is only needed when there is
+    // no requested address.
+    if (!is_aligned(base, alignment)) {
+      // Base not aligned, retry.
+      unmap_or_release_memory(base, size, fd != -1 /*is_file_mapped*/);
+      // Map using the requested alignment.
+      base = map_or_reserve_memory_aligned(size, alignment, fd, exec);
+    }
+  }
+
+  return base;
+}
+
+static char* reserve_memory_special(char* requested_address, const size_t size,
+                                    const size_t alignment,  bool exec) {
+
+  log_trace(pagesize)("Attempt special mapping: size: " SIZE_FORMAT "%s, "
+                      "alignment: " SIZE_FORMAT "%s",
+                      byte_size_in_exact_unit(size), exact_unit_for_byte_size(size),
+                      byte_size_in_exact_unit(alignment), exact_unit_for_byte_size(alignment));
+
+  char* base = os::reserve_memory_special(size, alignment, requested_address, exec);
+  if (base != NULL) {
+    // Check alignment constraints.
+    assert(is_aligned(base, alignment),
+           "reserve_memory_special() returned an unaligned address, base: " PTR_FORMAT
+           " alignment: " SIZE_FORMAT_HEX,
+           p2i(base), alignment);
+  } else {
+   if (large_pages_requested()) {
+      log_debug(gc, heap, coops)("Reserve regular memory without large pages");
+    }
```

Comment:
I think we should move this entire block to move to the caller.

Reply:
I thought about this as well, but wanted to keep `reserve(...)` as "clean" as possible and this log-message is connected to reserving large pages which is handled in `reserve_memory_special(...)`. 

The strange thing is that this log is using the tag-set `gc, heap, coops` (pre-existing) which suggests that we could move it all the way out to `try_reserve_heap(...)`. The problem with doing that is that this log is used by the test: TestLargePageUseForAuxMemory for other mappings than the heap. It would most likely make sense to move this message to a different tag-set or possibly split it so that this log only is printed for the heap and we have a different one for other mappings. The sad part about that is that the log message and the "failure" becomes more separated.

Reply:
ok, that seems reasonable. We can leave it as is.